cloud_storage/cloud_storage/overrides/file.py

```diff
@@ -50,17 +52,74 @@ def on_trash(self) -> None:
 				_("This file is attached to a submitted document and cannot be deleted"),
 				frappe.PermissionError,
 			)
-
+		if len(self.file_association) > 1:
+			frappe.throw(
+				_("This file is attached to multiple documents and cannot be deleted"),
+				frappe.PermissionError,
+			)
 		super().on_trash()
 
+	def associate_files(self) -> File:
+		if not self.content_hash and "/api/method/retrieve" in self.file_url:  # type: ignore
+			associated_doc = frappe.get_value("File", {"file_url": self.file_url}, "name")
+		else:
+			associated_doc = frappe.db.get_value(
+				"File",
+				{"content_hash": self.content_hash, "name": ["!=", self.name], "is_folder": False} # type: ignore
+			)
+
+		if associated_doc:
+			existing_file = frappe.get_doc("File", associated_doc)
+			existing_file.attached_to_doctype = self.attached_to_doctype
+			existing_file.attached_to_name = self.attached_to_name
+			self.content_hash = existing_file.content_hash
+			self = existing_file
+
+		existing_attachment = list(
+			filter(
+				lambda row: row.link_doctype == self.attached_to_doctype
+				and row.link_name == self.attached_to_name,
+				self.file_association,
+			)
+		)
+		if not existing_attachment:
+			self.append(
+				"file_association",
+				{"link_doctype": self.attached_to_doctype, "link_name": self.attached_to_name},
+			)
```

Comment:
If there is no `associated_doc`, this block will not do anything

Reply:
Are you asking for a guard clause here to prevent execution?

Reply:
Re-read this and it should stay as-is; this prevents a second association of a file (attachment) with the File (doctype)